domains/client/domain-executor/src/aux_schema.rs

```diff
@@ -109,6 +128,11 @@ where
                 execution_receipt_key(primary_hash).as_slice(),
```

Comment:
The advantage of using primary block hash as the key has seemingly been diminished by v2 as now each primary block hash is mapped to an optional ER as well, let's reassess these two options. @NingLin-P Perhaps you could make a brief doc for them on the use cases so that we don't have to look into the code? I can do it as well if you prefer.

Reply:
> Perhaps you could make a brief doc for them on the use cases so that we don't have to look into the code?

I'm willing to do that, but I think there is not really a trade-off we need to make, using `domain_hash` will just bring back the issue https://github.com/subspace/subspace/issues/1254 because:

When processing the consensus block, the operator checks the ER contains inside is the same as the one it produces locally, if not then use the local ER to construct and submit a fraud proof.

When using `domain_hash` as key we are unable to find and load the local ER if the incoming ER is different from the local one because the `incoming_ER.domain_hash` is also different, thus `load_local_ER(incoming_ER.domain_hash)` will always result in `None` then we are unable to construct the fraud proof properly.

Reply:
The same problem is back, but there might be another solution given the context has been changed a lot, I'm not entirely sure, hence want to make it assessed again under the v2 design.

Reply:
I have revisited the v2 fraud proof, which only mentioned the operator should "detecting any discrepancies" but didn't give more detail.

But IMO, the requirement is still the same under v2 design, namely **get the local ER that driving from the same consensus block as the incoming ER** thus seems inevitable to use the consensus block hash as the key for ER.

Reply:
I meant to have a brief intro for the scenarios here so that we can come to a conclusion naturally and even help the readers who are not familiar with the context.

So, the problem here is how to track the receipt on the operator's side so that it can be retrieved later.

#### Use cases

1. Retrieve the receipt for bundle production
    - Given the tip receipt on the consensus chain, include the next receipt or confirm the tip receipt in the bundle.
    - Given domain_block_number ⇒ domain_block_hash ⇒ load the corresponding local receipt
    
2. Retrieve the receipt of a specific domain block in order to compare the local trace and the other trace for fraud proof.
    - Given an external receipt ⇒ external_receipt.consensus_block_hash ⇒ load the corresponding local receipt.
    - There is no `domain_block_hash` in the [v2 ExecutionReceipt structure](https://www.notion.so/subspacelabs/Domains-v2-Specification-WIP-3fb0ec6e4d204c4881a7df50ef58da8f?pvs=4#a11d80f571cd45e4a188e74d1de23e3d).
        Actually, even if we include `domain_block_hash` in the receipt, it’s unusable in case the bad receipt, because it’s never verified and any trace mismatch will cause a different domain block hash since the state must be different, `bad_receipt.domain_block_hash` never exists on the honest operator’s end.

#### Options

- Use `domain_block_hash` as the key (domain_block_hash ⇒ ER)
    - Works for use case 1 directly, but has no way to work for use case 2.
- Use `consensus_block_hash` as the key (consensus_block_hash ⇒ ER)
    - Works for use case 2 directly, but we have to additionally track a mapping of domain_block_hash ⇒ consensus_block_hash for use case 1.

To conclude, the receipt needs to be keyed by `consensus_block_hash`.